### PR TITLE
UCP/TAG: cleanup unexp queue on EP destroy

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
## What
cleanup unexp queue on EP destroy

## How ?
on top of https://github.com/yosefe/ucx/commit/2e71b375fa0b6fa5e86083df104808dcb3d51e12

### CODE REVIEW

**Comment by evgeny-leksikov** [comment]
@yosefe could you pls review

File: src/ucp/core/ucp_ep.c [hunk 1]
```diff
@@ -1072,9 +1154,11 @@ void ucp_ep_set_failed(ucp_ep_h ucp_ep, ucp_lane_index_t lane,
     }
 
     /* The EP can be closed from last completion callback */
+    ep_id = ucp_ep_local_id(ucp_ep);
     ucp_ep_discard_lanes(ucp_ep, status);
     ucp_ep_reqs_purge(ucp_ep, status);
     ucp_stream_ep_cleanup(ucp_ep, status);
+    ucp_tm_ep_cleanup(ucp_ep, ep_id);
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: why need to pass id? can't ucp_tm_ep_cleanup look it up?

File: src/ucp/core/ucp_ep.c [hunk 2]
```diff
@@ -1046,11 +1054,85 @@ static void ucp_ep_set_lanes_failed(ucp_ep_h ep, uct_ep_h *uct_eps)
...
+                continue;
+            }
+        } else {
+            eager_hdr = (const void*)(rdesc + 1);
+            if (eager_hdr->ep_id != ep_id) {
+                /* eager not matched */
+                continue;
+            }
+        }
+        ucs_debug("releasing unexpected rdesc %p", rdesc);
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: 1. space line before
2. print ep and its id

File: src/ucp/core/ucp_ep.c [hunk 3]
```diff
@@ -1046,11 +1054,85 @@ static void ucp_ep_set_lanes_failed(ucp_ep_h ep, uct_ep_h *uct_eps)
...
+        ucp_tag_unexp_remove(rdesc);
+        ucp_recv_desc_release(rdesc);
+    }
+    /* remove from fragments hash */
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:1103]: space line before

File: src/ucp/core/ucp_ep.c [hunk 4]
```diff
@@ -1046,11 +1054,85 @@ static void ucp_ep_set_lanes_failed(ucp_ep_h ep, uct_ep_h *uct_eps)
...
+    kh_foreach_key(&tm->frag_hash, msg_id, {
+        iter   = kh_get(ucp_tag_frag_hash, &tm->frag_hash, msg_id);
+        matchq = &kh_val(&tm->frag_hash, iter);
+        if (!ucp_tag_frag_match_is_unexp(matchq)) {
+            /* remove receive request from expected hash */
+            rreq = matchq->exp_req;
+            if (rreq->recv.tag.ep_id == ep_id) {
+                ucs_debug("completing req %p", rreq);
+                ucp_request_complete_tag_recv(rreq, UCS_ERR_CANCELED);
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: pass status to this function instead of always using CANCELED

File: src/ucp/core/ucp_ep.c [hunk 5]
```diff
@@ -1046,11 +1054,85 @@ static void ucp_ep_set_lanes_failed(ucp_ep_h ep, uct_ep_h *uct_eps)
...
+            rdesc = ucs_queue_head_elem_non_empty(&matchq->unexp_q, ucp_recv_desc_t,
+                                                  tag_frag_queue);
+            ucs_assert(!(rdesc->flags & UCP_RECV_DESC_FLAG_RNDV));
+            eager_mid_hdr = (void*)(rdesc + 1);
+            if (eager_mid_hdr->ep_id != ep_id) {
+                continue;
+            }
+
+            ucs_queue_for_each_extract(rdesc, &matchq->unexp_q, tag_frag_queue, 1) {
+                ucs_debug("releasing unexpected rdesc %p", rdesc);
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: print ep

File: src/ucp/core/ucp_ep.c [hunk 6]
```diff
@@ -1046,11 +1054,85 @@ static void ucp_ep_set_lanes_failed(ucp_ep_h ep, uct_ep_h *uct_eps)
...
+    }
+    /* remove from fragments hash */
+    kh_foreach_key(&tm->frag_hash, msg_id, {
+        iter   = kh_get(ucp_tag_frag_hash, &tm->frag_hash, msg_id);
+        matchq = &kh_val(&tm->frag_hash, iter);
+        if (!ucp_tag_frag_match_is_unexp(matchq)) {
+            /* remove receive request from expected hash */
+            rreq = matchq->exp_req;
+            if (rreq->recv.tag.ep_id == ep_id) {
+                ucs_debug("completing req %p", rreq);
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: ucs_debug("ep %p: completing expected receive request %p with status %s", ep, req)

File: src/ucp/tag/tag_match.inl [hunk 7]
```diff
@@ -145,16 +145,17 @@ ucp_tag_unexp_remove(ucp_recv_desc_t *rdesc)
...
 {
     ucs_list_link_t *hash_list;
 
     hash_list = ucp_tag_unexp_get_list_for_tag(tm, tag);
     ucs_list_add_tail(hash_list,           &rdesc->tag_list[UCP_RDESC_HASH_LIST]);
     ucs_list_add_tail(&tm->unexpected.all, &rdesc->tag_list[UCP_RDESC_ALL_LIST]);
 
-    ucs_trace_req("unexp "UCP_RECV_DESC_FMT" tag %"PRIx64,
-                  UCP_RECV_DESC_ARG(rdesc), tag);
+    ucs_trace_req("unexp "UCP_RECV_DESC_FMT" tag %"PRIx64" ep %"PRIx64,
```
Comment (yosefe) [inline src/ucp/tag/tag_match.inl]: rename ep -> ep_id

File: src/ucp/core/ucp_request.h [hunk 8]
```diff
@@ -348,6 +348,7 @@ struct ucp_request {
                     ucp_tag_t                   tag;        /* Expected tag */
                     ucp_tag_t                   tag_mask;   /* Expected tag mask */
                     uint64_t                    sn;         /* Tag match sequence */
+                    ucs_ptr_map_key_t           ep_id;      /* Endpoint local id */
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h:351]: need to reset it to KEY_INVALID when first creating the request - in tag_recv_common

File: src/ucp/core/ucp_ep.c (hunk 1 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.c]: `ucp_ep_discard_lanes` above invalidates it, it's used for data drop on failed EP

Comment (yosefe) [reply src/ucp/core/ucp_ep.c]: ok, maybe worth to add comment

File: src/ucp/core/ucp_request.h (hunk 8 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_request.h:351]: done

**Review by yosefe — APPROVED** [review]

File: src/ucp/core/ucp_ep.c (hunk 1 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.c]: added

**Comment by evgeny-leksikov** [comment]
@yosefe ok to merge?

**Pull request merged**